Interactions, Slash-Commands, and Buttons

MinnDevelopment opened the pull request:
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

<!--
  There are several guidelines you should follow in order for your
  Pull Request to be merged.
-->

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

<!--
  It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.
-->

### Changes

- [x] Internal code
- [x] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ <!-- Insert other type here -->

<!-- Replace "NaN" with an issue number if this is a response to an issue -->

Closes Issue: NaN

## Description

This pull request adds support for [Slash Commands](https://discord.com/developers/docs/interactions/slash-commands) and webhook execution.

Example: [SlashBotExample](https://github.com/DV8FromTheWorld/JDA/blob/feature/slash-commands/src/examples/java/SlashBotExample.java)
Documentation: https://javadoc.jitpack.io/com/github/dv8fromtheworld/jda/bf71f22/javadoc/index.html
Guide: https://github.com/DV8FromTheWorld/JDA/wiki/Interactions (wip)

## How can I test a preview of this?

Add this branch as a dependency through jitpack: [How?](https://github.com/DV8FromTheWorld/JDA/wiki/10%29-FAQ#how-do-i-test-other-builds)

The snapshot would be `com.github.dv8fromtheworld:jda:commit` from jitpack. (replace `commit` with the latest commit hash in this PR)

In order to create guild commands you need the `applications.commands` scope authorized for that guild. So instead of `scope=bot` in your authorization URL, you must have `scope=bot+applications.commands`.

It is highly recommended to use guild-specific slash commands for testing as they can be updated a lot more frequently than global commands. Create a test guild and invite your bot with the aforementioned scopes, then use Guild#updateCommands.

## [What is possible?](https://gist.github.com/MinnDevelopment/59af01710f404ef58827b56de96fb9e9)

# TODO

- [x] SlashCommandEvent and InteractionCreateEvent
- [x] Creating Commands (global/guild)
- [x] Deleting Commands (global/guild)
- [x] Editing Commands (global/guild)
- [x] Ephemeral messages
- [x] Interaction webhooks
- [x] Application Command Create/Update/Delete events
- [x] [Slash Command Permissions](https://github.com/discord/discord-api-docs/pull/2737)
- [x] [Components aka Buttons](https://github.com/DV8FromTheWorld/JDA/pull/1612)
- [x] Example (See [SlashBotExample](https://github.com/DV8FromTheWorld/JDA/blob/feature/slash-commands/src/examples/java/SlashBotExample.java))
- [x] Documentation
- [x] Checks
- [ ] ~~Outgoing Webhook support (interactions through webservers)~~ Putting this on hold for now

MinnDevelopment commented:
This is still in very early stages and needs a lot more polishing, just posting this to show some progress. I'm still very busy with other projects besides JDA for another month or so.

MinnDevelopment commented:
Here is my current testing code in kotlin which abuses MethodHandles to provide a simpler implementation:

```kotlin
// This is just a helper class to allow using method names as commands
open class CommandManager : EventListener {
    private val lookup = MethodHandles.lookup()
    private val methods = mutableMapOf<String, MethodHandle>()
    private val type = MethodType.methodType(Void.TYPE, SlashCommandEvent::class.java)

    override fun onEvent(event: GenericEvent) {
        if (event !is SlashCommandEvent)
            return

        try {
            val name = event.name.toLowerCase()
            val method = methods.computeIfAbsent(name) {
                lookup.findVirtual(this.javaClass, name, type)
            }

            method.invokeWithArguments(this, event)
        } catch (ignored: NoSuchMethodException) {}
    }
}


class Moderation : CommandManager() {
    fun ping(event: SlashCommandEvent) { // function name = command name
        event.thread.editOriginal("Test followup 1").setEphemeral(true).queue()
        event.thread.editOriginal("Test followup 2").setEphemeral(true).queue()
        event.thread.editOriginal("Test followup 3").setEphemeral(true).queue()
        event.thread.editOriginal("Test followup 4").setEphemeral(true).queue()
        event.thread.editOriginal("Test followup 5").setEphemeral(true).queue()
        event.thread.editOriginal("Test followup 6").setEphemeral(true).queue() // This is actually rate limited (major parameter is the APPLICATION ID here, pretty insane imo)
        event.reply("Test reply").setEphemeral(true).queue() // You can queue up followups before even acknowledging the command! They just execute after, so there is no need for callbacks!
    }

    fun kick(event: SlashCommandEvent) {
        val userOption = event.getOptionsByName("user").first()
        // Can be resolved to either user
        event.reply("This would kick ${userOption.asUser?.asTag}! [Button to accept would be here]").setEphemeral(true).queue()

        // Or to member! It will just be null if the user isn't a member of the server...
        userOption.asMember?.kick()?.queue {
            event.thread.sendMessage("User was kicked successfully!").setEphemeral(true).queue() // This will send *after* the previous callback completes
        }
    }
}
```

As you can see, the callbacks are not necessary anymore due to the way I delay followup calls with the new `TriggerRestAction` that can be triggered later.

MinnDevelopment commented:
I'm adding `interaction_token` as a new major parameter. The follow-up webhook endpoints use the interaction id internally to identify the rate limit bucket. This id is not present in the endpoint itself so we use the token instead.

In order to keep logs clean, I use `String#hashCode` to calculate a relatively unique integer instead. Since these tokens are not generated by user input they can use the simple hash function provided by `String#hashCode` rather than a more expensive cryptographic alternative. The raw `interaction_token` is a very long string that would make the logs unreadable and fill them with useless information. In my testing, that token had up to 212 characters. (>.>)

Justification: discord/discord-api-docs#2701

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/Webhook.java:
```diff
@@ -36,8 +38,10 @@
-public interface Webhook extends ISnowflake, IFakeable
+public interface Webhook extends ISnowflake, WebhookClient<WebhookMessageAction>, IFakeable
 {
+    Pattern WEBHOOK_URL = Pattern.compile("https?://(?:[^\\s.]+\\.)discord(?:app)\\.com/api(?:/v\\d+)?/webhooks/(\\d+)/([^\\s/]+)", Pattern.CASE_INSENSITIVE);
```
```suggestion
    Pattern WEBHOOK_URL = Pattern.compile("https?://(?:[^\\s.]+\\.)?discord(?:app)\\.com/api(?:/v\\d+)?/webhooks/(\\d+)/([^\\s/]+)", Pattern.CASE_INSENSITIVE);
```

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/Webhook.java:
```diff
@@ -36,8 +38,10 @@
-public interface Webhook extends ISnowflake, IFakeable
+public interface Webhook extends ISnowflake, WebhookClient<WebhookMessageAction>, IFakeable
 {
+    Pattern WEBHOOK_URL = Pattern.compile("https?://(?:[^\\s.]+\\.)?discord(?:app)\\.com/api(?:/v\\d+)?/webhooks/(\\d+)/([^\\s/]+)", Pattern.CASE_INSENSITIVE);
```
```suggestion
    Pattern WEBHOOK_URL = Pattern.compile("https?://(?:[^\\s.]+\\.)?discord(?:app)?\\.com/api(?:/v\\d+)?/webhooks/(\\d+)/([^\\s/]+)", Pattern.CASE_INSENSITIVE);
```

duncte123 commented on src/main/java/net/dv8tion/jda/api/JDA.java:
```diff
@@ -492,6 +489,47 @@ default JDA awaitReady() throws InterruptedException
+
+    @Nonnull
+    @CheckReturnValue
+    CommandCreateAction upsertCommand(@Nonnull String name, @Nonnull String description);
```
Maybe add `CommandCreateAction upsertCommand(@Nonnull CommandData command);` for convenience

MinnDevelopment commented:
Buttons are now available on all guilds: https://support.discord.com/hc/en-us/articles/1500012250861-Bots-Buttons

duncte123 commented:
The message builder constructor currently does not copy the action rows. Is this intended?

MinnDevelopment commented:
> also can't seem clear the embed when editing via deferEdit

The correct way is doing `setEmbeds(Collections.emptyList())`. If that doesn't work, it's still a discord bug at this time.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/SelfUser.java:
```diff
@@ -31,6 +31,14 @@
+    long getApplicationIdLong();
+
+    @Nonnull
+    default String getApplicationId()
```
Missing documentation

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/Webhook.java:
```diff
@@ -38,6 +39,8 @@
  */
 public interface Webhook extends ISnowflake, IFakeable
 {
+    Pattern WEBHOOK_URL = Pattern.compile("https?://(?:[^\\s.]+\\.)?discord(?:app)?\\.com/api(?:/v\\d+)?/webhooks/(\\d+)/([^\\s/]+)", Pattern.CASE_INSENSITIVE);
```
Missing documentation

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/application/ApplicationCommandCreateEvent.java:
```diff
@@ -0,0 +1,32 @@
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public class ApplicationCommandCreateEvent extends GenericApplicationCommandEvent
```
Missing documentation

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/application/ApplicationCommandDeleteEvent.java:
```diff
@@ -0,0 +1,32 @@
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public class ApplicationCommandDeleteEvent extends GenericApplicationCommandEvent
```
Missing documentation

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/application/ApplicationCommandUpdateEvent.java:
```diff
@@ -0,0 +1,32 @@
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public class ApplicationCommandUpdateEvent extends GenericApplicationCommandEvent
```
Missing documentation

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/application/GenericApplicationCommandEvent.java:
```diff
@@ -0,0 +1,50 @@
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public abstract class GenericApplicationCommandEvent extends Event
```
Missing documentation

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/interactions/ActionRow.java:
```diff
@@ -0,0 +1,148 @@
+
+    @Nonnull
+    @Override
+    public List<Button> getButtons()
```
Missing documentation

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/requests/GatewayIntent.java:
```diff
@@ -142,7 +142,9 @@
      */
-    DIRECT_MESSAGE_TYPING(14);
+    DIRECT_MESSAGE_TYPING(14),
+//    INTERACTION_CREATE(15)
```
Remove this

MinnDevelopment commented:
The complete documentation can be viewed on jitpack: https://javadoc.jitpack.io/com/github/dv8fromtheworld/jda/1b8f559/javadoc/index.html

MinnDevelopment commented:
I've implemented support for command permissions in the most recent commit. I call them `CommandPrivilege` here to avoid confusion with the similar named `Permission` system used for roles and overwrites.

I haven't really tested the implementation of this much since I think permissions are bad and annoying to use, but if you find any issues please let me know!

WalshyDev commented:
It appears the button gets removed if you do `Message#editMessage(MessageEmbed)` but not if you do `Message#editMessage(Message)`

Considering `content` persists in editMessage(MessageEmbed) I would assume that buttons do (and you can't add buttons to MessageEmbed)

```java
/* - Button is kept in this code
final Message sent = e.getChannel().sendMessage(new MessageBuilder()
    .setContent("123")
    .setActionRows(ActionRow.of(
        Button.primary("test", "Testing 123")
    ))
    .build()
).complete();

sent.editMessage(new MessageBuilder(sent).setContent("456").build()).complete();
 */

// Button is removed in this code
final Message sending = new MessageBuilder()
    .setContent("content")
    .setEmbed(new EmbedBuilder().setDescription("123").build())
    .setActionRows(ActionRow.of(
        Button.primary("test", "Testing 123")
    ))
    .build();
final Message sent = e.getChannel().sendMessage(sending).complete();

sent.editMessage(new EmbedBuilder().setDescription("456").build()).complete();

e.reply("did it remove the button?").queue();
```

MinnDevelopment commented:
@WalshyDev should be fixed in a2ad487

WalshyDev commented:
Confirmed it's fixed, appreciate the quick patch :+1:

caneleex commented on src/main/java/net/dv8tion/jda/api/interactions/Interaction.java:
```diff
@@ -0,0 +1,471 @@
+    /**
+     * Reply to this interaction and acknowledge it.
+     * <br>This will send a reply message for this interaction.
+     * You can use {@link ReplyAction#setEphemeral(boolean) setEohemeral(true)} to only let the target user see the message.
```
```suggestion
     * You can use {@link ReplyAction#setEphemeral(boolean) setEphemeral(true)} to only let the target user see the message.
```

VixenKasai commented:
Can we overload the OptionData constructor to include the requirement parameter to help cut down on clutter when using varargs?

```
public OptionData(@Nonnull OptionType type, @Nonnull String name, @Nonnull String description, @Nonnull boolean isRequired)    { 
       Checks.notNull(type, "Type"); 
       Checks.notEmpty(name, "Name");
       Checks.notEmpty(description, "Description");
       Checks.notLonger(name, 32, "Name");
       Checks.notLonger(description, 100, "Description");
       Checks.matches(name, Checks.ALPHANUMERIC_WITH_DASH, "Name"); 
       Checks.isLowercase(name, "Name");
       this.type = type;
       this.name = name;
       this.description = description;
       this.isRequired = isRequired;
       if (type.canSupportChoices())  choices = new LinkedHashMap&lt;&gt;();
}
```

caneleex commented on src/main/java/net/dv8tion/jda/api/interactions/Interaction.java:
```diff
@@ -0,0 +1,471 @@
+    /**
+     * Reply to this interaction and acknowledge it.
+     * <br>This will send a reply message for this interaction.
+     * You can use {@link ReplyAction#setEphemeral(boolean) setEohemeral(true)} to only let the target user see the message.
```
```suggestion
     * You can use {@link ReplyAction#setEphemeral(boolean) setEphemeral(true)} to only let the target user see the message.
```

caneleex commented on src/main/java/net/dv8tion/jda/api/interactions/Interaction.java:
```diff
@@ -0,0 +1,471 @@
+    /**
+     * Reply to this interaction and acknowledge it.
+     * <br>This will send a reply message for this interaction.
+     * You can use {@link ReplyAction#setEphemeral(boolean) setEohemeral(true)} to only let the target user see the message.
```
```suggestion
     * You can use {@link ReplyAction#setEphemeral(boolean) setEphemeral(true)} to only let the target user see the message.
```

caneleex commented on src/main/java/net/dv8tion/jda/api/interactions/Interaction.java:
```diff
@@ -0,0 +1,471 @@
+//    /**
+//     * Reply to this interaction and acknowledge it.
+//     * <br>This will send a reply message for this interaction.
+//     * You can use {@link ReplyAction#setEphemeral(boolean) setEohemeral(true)} to only let the target user see the message.
```
```suggestion
//     * You can use {@link ReplyAction#setEphemeral(boolean) setEphemeral(true)} to only let the target user see the message.
```

caneleex commented on src/main/java/net/dv8tion/jda/api/interactions/Interaction.java:
```diff
@@ -0,0 +1,471 @@
+    /**
+     * Reply to this interaction and acknowledge it.
+     * <br>This will send a reply message for this interaction.
+     * You can use {@link ReplyAction#setEphemeral(boolean) setEohemeral(true)} to only let the target user see the message.
```
```suggestion
     * You can use {@link ReplyAction#setEphemeral(boolean) setEphemeral(true)} to only let the target user see the message.
```

Wirlie commented:
Hello, the Command class are not mapping groups, subcommands and options correctly, this is my testing code:

```java
jda.upsertCommand(
    CommandData("test", "A test command")
        .addSubcommands(
            SubcommandData("testsubcommand", "A test subcommand")
                .addOption(OptionType.STRING, "text", "A text option")
        )
).complete()

for(command in jda.retrieveCommands().complete()) {
    println("Command: ${command.name}")
    for(subCommandGroup in command.subcommandGroups) {
        println("|--> SubCommandGroup: ${subCommandGroup.name}")
        for(subCommand in subCommandGroup.subcommands) {
            println("     |--> SubCommand: ${subCommand.name}")
            for(subOption in subCommand.options) {
                println("           |--> Option: ${subOption.name} | TYPE=${subOption.type}")
            }
        }
    }
    for(subCommand in command.subcommands) {
        println("|--> SubCommand: ${subCommand.name}")
        for(subOption in subCommand.options) {
            println("     |--> Option: ${subOption.name} | TYPE=${subOption.type}")
        }
    }
    for(option in command.options) {
        println("|--> Option: ${option.name} | TYPE=${option.type}")
    }
}
```

This is the printed message:
```
Command: test
|--> SubCommandGroup: testsubcommand
     |--> SubCommand: text
|--> SubCommand: testsubcommand
     |--> Option: text | TYPE=STRING
|--> Option: testsubcommand | TYPE=SUB_COMMAND
```
Probably the expected message would be:
```
Command: test
|--> SubCommand: testsubcommand
     |--> Option: text | TYPE=STRING
```

MinnDevelopment commented:
@Wirlie should be fixed in 757de9b

duncte123 commented:
Is there any info on component type 3 (drop-down)?

MinnDevelopment commented:
Selection menus (type 3) are currently in private beta and will not be part of this pull request.

Chew commented:
Some notes from my testing:
- Only 10 privileges may be added (at a time?). Got a BASE_TYPE_MAX_LENGTH must be 10 or fewer in length.
- `APPLICATION_COMMAND_PERMISSIONS_UPDATE` still seems to be unhandled
- I could not, for the life of me, get `CommandPrivilege.enableUser(String userId)` to work. Added my User ID and still couldn't run the command. Added a role I had and it worked fine.
- Can we get a way to clear commands in a Guild? It's undocumented, but if you `PUT api/v8/applications/{applicationId}/guilds/{guildId}/commands` it clears all commands.

Xirado commented:
> Can we get a way to clear commands in a Guild? It's undocumented, but if you PUT api/v8/applications/{applicationId}/guilds/{guildId}/commands with a body of [] it clears all commands.

I think this is possible when queueing an empty CommandUpdateAction

MinnDevelopment commented:
> I could not, for the life of me, get CommandPrivilege.enableUser(String userId) to work. Added my User ID and still couldn't run the command. Added a role I had and it worked fine.

Works fine for me:

```kotlin
    val guild = jda.getGuildById(guildId)
    guild.upsertCommand("test", "Disabled?")
           .setDefaultEnabled(false)
           .flatMap {
               guild.updateCommandPrivileges(it.id, listOf(enableUser(86699011792191488)))
           }
           .queue()
```

![image](https://user-images.githubusercontent.com/18090140/120069185-d19ee680-c084-11eb-99c9-cd0fd6a02648.png)

MinnDevelopment commented:
> APPLICATION_COMMAND_PERMISSIONS_UPDATE still seems to be unhandled

That event seems to be undocumented so I won't support it.

duncte123 commented:
> > I could not, for the life of me, get CommandPrivilege.enableUser(String userId) to work. Added my User ID and still couldn't run the command. Added a role I had and it worked fine.
> 
> Works fine for me:
> 
> ```kotlin
>     val guild = jda.getGuildById(guildId)!!
>     guild.upsertCommand("test", "Disabled?")
>            .setDefaultEnabled(false)
>            .flatMap {
>                guild.updateCommandPrivileges(it.id, listOf(enableUser(86699011792191488)))
>            }
>            .queue()
> ```
> 
> ![image](https://user-images.githubusercontent.com/18090140/120069185-d19ee680-c084-11eb-99c9-cd0fd6a02648.png)

Is it possible to send this along with the command data so we can chain it?

duncte123 commented:
The following code produces an exception.
```kotlin
guild.upsertCommand(
        CommandData("test", "A test command")
            .setDefaultEnabled(false)
    )
    .flatMap {
        it.delete()
    }
    .queue()
```

```
09:53:48.704 ForkJoinPool.commonPool-worker-3   RestAction      ERROR  Encountered error while processing success consumer
java.lang.IllegalArgumentException: Error Compiling Route: [applications/{application_id}/guilds/{guild_id}/commands/{command_id}], incorrect amount of parameters provided.Expected: 3, Provided: 2
	at net.dv8tion.jda.internal.requests.Route.compile(Route.java:336)
	at net.dv8tion.jda.api.interactions.commands.Command.delete(Command.java:103)
	at me.duncte123.eduard.listeners.Listener$onEvent$1.apply(Listener.kt:61)
	at me.duncte123.eduard.listeners.Listener$onEvent$1.apply(Listener.kt:33)
	at net.dv8tion.jda.internal.requests.restaction.operator.FlatMapRestAction.supply(FlatMapRestAction.java:44)
	at net.dv8tion.jda.internal.requests.restaction.operator.FlatMapRestAction.lambda$queue$0(FlatMapRestAction.java:54)
	at net.dv8tion.jda.api.requests.Request.lambda$onSuccess$0(Request.java:95)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1429)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

MinnDevelopment commented:
> Is it possible to send this along with the command data so we can chain it?

No.

duncte123 commented:
Could you add OptionData#addOptionData or something similar?
Otherwise setting an option data with the type of `SUB_COMMAND` would be completely useless

MinnDevelopment commented:
> Otherwise setting an option data with the type of SUB_COMMAND would be completely useless

That is intentional, you should use addSubcommands instead.

cbyrneee commented:
I can't seem to be able to remove the action row from a message. If I want to remove all buttons, whilst editing my message, how would I go about doing that?

![image](https://user-images.githubusercontent.com/71222289/120114918-4d7a5b00-c179-11eb-8162-b4b751849ca8.png)
![image](https://user-images.githubusercontent.com/71222289/120114933-5c610d80-c179-11eb-95b6-d59f14e89ef7.png)

MinnDevelopment commented:
You should use `setActionsRows(Collections.emptyList())` @cbyrneee

cbyrneee commented:
> 
> 
> You should use `setActionsRows(Collections.emptyList())` @cbyrneee

Awesome that works, thank you for the response!

DV8FromTheWorld commented on src/main/java/net/dv8tion/jda/internal/requests/restaction/WebhookMessageActionImpl.java:
```diff
@@ -0,0 +1,253 @@
+        return this;
+    }
+
+    private DataObject getJSON()
```
Lets change this to `toData`

DV8FromTheWorld commented on src/main/java/net/dv8tion/jda/internal/requests/restaction/CommandEditActionImpl.java:
```diff
@@ -0,0 +1,196 @@
+public class CommandEditActionImpl extends RestActionImpl<Command> implements CommandEditAction
+{
+    private static final String UNDEFINED = "undefined";
+    private static final int NAME_SET = 1, DESCRIPTION_SET = 2, OPTIONS_SET = 4;
```
Lets defined these on separate lines with proper bit-shifts like we do in `WebhookMessageUpdateActionImpl`

DV8FromTheWorld commented on src/main/java/net/dv8tion/jda/internal/requests/restaction/interactions/CallbackActionImpl.java:
```diff
@@ -0,0 +1,131 @@
+        this.hook = hook;
+    }
+
+    protected abstract DataObject getJSON();
```
Lets name this `toData` to match other places.

DV8FromTheWorld commented on src/main/java/net/dv8tion/jda/internal/requests/restaction/MessageActionImpl.java:
```diff
@@ -519,6 +468,10 @@ protected DataObject getJSON()
                 obj.putNull("nonce");
             else
                 obj.put("nonce", nonce);
+            if (components == null)
```
It wasn't setup in this PR, but this method should really be called `toData`

MinnDevelopment replied:
Out of scope

DV8FromTheWorld commented on src/main/java/net/dv8tion/jda/internal/requests/restaction/WebhookMessageActionImpl.java:
```diff
@@ -0,0 +1,253 @@
+    private DataObject getJSON()
+    {
+        DataObject json = DataObject.empty();
+        json.put("content", content.toString());
```
Is it possible that this is why that `editOriginalComponents` doesn't work? We're always sending a string, even when it is empty (`""`)

For `MessageActionImpl` when `content.length() == 0` we `putNull("content")`

MinnDevelopment replied:
That is only because MessageActionImpl deals with edits. This class is not for edits.

DV8FromTheWorld commented on src/main/java/net/dv8tion/jda/internal/requests/restaction/interactions/ReplyActionImpl.java:
```diff
@@ -0,0 +1,233 @@
+    public ReplyActionImpl setEphemeral(boolean ephemeral)
+    {
+        if (ephemeral)
+            this.flags |= 64;
```
Where is this `64` coming from?

MinnDevelopment replied:
Docs.

DV8FromTheWorld commented on src/main/java/net/dv8tion/jda/internal/requests/restaction/interactions/UpdateActionImpl.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @Override
+    protected DataObject getJSON()
```
Switch `toData`?

DV8FromTheWorld reviewed (commented):
What a monumental effort here. Nice work.

arynxd commented on src/main/java/net/dv8tion/jda/api/events/interaction/ButtonClickEvent.java:
```diff
@@ -0,0 +1,104 @@
+ * To receive these events, you must unset the <b>Interactions Endpoint URL</b> in your application dashboard.
+ * You can simply remove the URL for this endpoint in your settings at the <a href="https://discord.com/developers/applications" target="_blank">Discord Developers Portal</a>.
+ */
+public class ButtonClickEvent extends GenericInteractionCreateEvent implements ButtonInteraction
```
Add isFromGuild to this

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/JDA.java:
```diff
@@ -492,6 +491,214 @@ default JDA awaitReady() throws InterruptedException
+     * <br>If a command with the same name exists, it will be replaced.
+     *
+     * <p>To specify a complete list of all commands you can use {@link #updateCommands()} instead.
+     *
+     * <p>You need the OAuth2 scope {@code "applications.commands"} in order to add commands to a guild.
```
You can't use global commands in a guild without this scope either.

DV8FromTheWorld replied:
Oh indeed? Interesting. Alrighty.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/Guild.java:
```diff
@@ -75,6 +75,317 @@
+     */
+    @Nonnull
+    @CheckReturnValue
+    RestAction<Void> updateCommandPrivileges(@Nonnull String id, @Nonnull Collection<? extends CommandPrivilege> privileges);
```
You can just do `Command#updatePrivileges` instead

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/exceptions/InteractionFailureException.java:
```diff
@@ -0,0 +1,35 @@
+ * <p>This is used to signal that a {@link net.dv8tion.jda.api.requests.restaction.WebhookMessageAction WebhookMessageAction}
+ * was cancelled due to a cascading failure from the initial command acknowledgement.
+ */
+public class InteractionFailureException extends CancellationException
```
I don't think thats needed

DV8FromTheWorld replied:
No big deal honestly. Just something that came to mind as I could see us having other InteractionFailures.

MinnDevelopment replied:
I think its better to have this name and we can just extend it anyway. That way if users handle interaction failures, they still handle all kinds of failures even when we introduce more concrete ones.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/interactions/Interaction.java:
```diff
@@ -0,0 +1,474 @@
+
+/**
+ * Abstract representation for any kind of Discord interaction.
+ * <br>This includes things such as {@link net.dv8tion.jda.api.interactions.commands.CommandInteraction Slash-Commands} or {@link ComponentInteraction Buttons}.
```
The ComponentsInteraction docs are a bit more detailed so I used that instead.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/interactions/commands/build/BaseCommand.java:
```diff
@@ -0,0 +1,134 @@
+
+    public BaseCommand(@Nonnull String name, @Nonnull String description)
+    {
+        Checks.notEmpty(name, "Name");
```
We don't need to since the alphanumerics check will fail

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/interactions/commands/build/OptionData.java:
```diff
@@ -0,0 +1,349 @@
+     * @see #addChoice(String, String)
+     */
+    @Nonnull
+    public List<Command.Choice> getChoices()
```
Because technically the API uses a list too which might cause issues if duplicate Choices are present for whatever reason.

DV8FromTheWorld replied:
I see. So the internal map is a JDA optimization and exposing as a List is guarding us against exposing an optimization that may need to be unwound in the future?

MinnDevelopment replied:
Yeah internals can be more optimized but the interface should be robust for the future instead

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/interactions/components/ActionRow.java:
```diff
@@ -0,0 +1,158 @@
+     * @return The action row
+     */
+    @Nonnull
+    public static ActionRow of(@Nonnull Component... components)
```
ActionRow does not implement Component so this is fine.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/interactions/components/Button.java:
```diff
@@ -0,0 +1,563 @@
+ *
+ * <p>Each button has either a {@code custom_id} or URL attached.
+ * The id has to be provided by the user and can be used to identify the button in the {@link net.dv8tion.jda.api.events.interaction.ButtonClickEvent ButtonClickEvent}.
+ *
```
That can be added later once I finish the wiki guide and upload the images to our assets branch

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/interactions/components/Button.java:
```diff
@@ -0,0 +1,563 @@
+     * @return The button instance
+     */
+    @Nonnull
+    static Button primary(@Nonnull String id, @Nonnull String label)
```
Reducing the number of factory methods. This was the simplest way to do that.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/requests/restaction/WebhookMessageAction.java:
```diff
@@ -0,0 +1,377 @@
+ *
+ * @see    net.dv8tion.jda.api.entities.WebhookClient#sendMessage(String)
+ */
+// TODO: WebhookMessage type (no channel/guild attached)
```
No this has been rescheduled to a later date as it seems out of scope for now. I introduced a generic type parameter so we can introduce this new type in the future.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/requests/restaction/WebhookMessageAction.java:
```diff
@@ -0,0 +1,377 @@
+     */
+    @Nonnull
+    @CheckReturnValue
+    WebhookMessageAction<T> setUsername(@Nullable String name);
```
Well it can be used it just does nothing

MinnDevelopment replied:
I commented them out for now

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/requests/restaction/WebhookMessageAction.java:
```diff
@@ -0,0 +1,377 @@
+     */
+    @Nonnull
+    @CheckReturnValue
+    WebhookMessageAction<T> setEphemeral(boolean ephemeral);
```
InteractionHook is currently the only implementation of this.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/requests/restaction/interactions/ReplyAction.java:
```diff
@@ -0,0 +1,237 @@
+//    @CheckReturnValue
+//    CommandReplyAction addFile(@Nonnull InputStream data, @Nonnull String name, @Nonnull AttachmentOption... options);
+
+//    enum Flag
```
Yes.

DV8FromTheWorld replied:
Lets clean up and remove the commented out code then until a time at which is needs to be introduced

MinnDevelopment commented on src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java:
```diff
@@ -1091,7 +1092,11 @@ public Message createMessage(DataObject jsonObject, @Nullable MessageChannel cha
     {
         long channelId = jsonObject.getUnsignedLong("channel_id");
         if (channel != null && channelId != channel.getIdLong())
-            channel = null;
+        {
+            channel = api.getTextChannelById(channelId);
+            if (channel == null)
+                channel = api.getPrivateChannelById(channelId);
+        }
```
For possible cross-channel replies that are currently not a thing but could break code in the future if discord decided to make a breaking change randomly.

DV8FromTheWorld replied:
Well alrighty then

MinnDevelopment commented on src/main/java/net/dv8tion/jda/internal/handle/InteractionCreateHandler.java:
```diff
@@ -0,0 +1,83 @@
+
+        switch (type)
+        {
+            case 2: // slash commands
```
I don't think introducing an enum for a single switch makes sense

MinnDevelopment commented on src/main/java/net/dv8tion/jda/internal/handle/InteractionCreateHandler.java:
```diff
@@ -0,0 +1,83 @@
+
+    private void handleAction(DataObject content)
+    {
+        if (content.getObject("data").getInt("component_type") != 2)
```
That switch will be introduced by the selection menu changes.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/internal/interactions/CommandInteractionImpl.java:
```diff
@@ -0,0 +1,166 @@
+        DataArray options = commandData.optArray("options").orElseGet(DataArray::empty);
+        DataObject resolveJson = commandData.optObject("resolved").orElseGet(DataObject::empty);
+
+        if (options.length() == 1)
```
You can't use multiple subcommands at the same time.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/internal/requests/restaction/MessageActionImpl.java:
```diff
@@ -51,46 +52,19 @@
     protected final StringBuilder content;
     protected final MessageChannel channel;
+    protected final AllowedMentionsUtil allowedMentions = new AllowedMentionsUtil();
+    protected List<ActionRow> components;
```
I want to stick to concrete types for the time being.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/requests/restaction/interactions/CallbackAction.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * A callback action is used to <b>acknowledge</b> an {@link net.dv8tion.jda.api.interactions.Interaction Interaction}.
+ */
+public interface CallbackAction extends RestAction<InteractionHook>
```
It is supposed to be generic, this is simply an abstraction to reduce code-duplication

DV8FromTheWorld replied:
I meant something like `InteractionCallbackAction`

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/utils/data/SerializableArray.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Allows custom serialization for JSON payloads of an object.
+ */
+public interface SerializableArray
```
It's used so you can do stuff like `put("array", new Foo())` which would be serialized to a DataArray.

DV8FromTheWorld replied:
Is there a reason we wouldn't want to do `SerializableObject` too?

MinnDevelopment replied:
We have that, its called SerializableData

MinnDevelopment commented on src/main/java/net/dv8tion/jda/internal/entities/WebhookImpl.java:
```diff
@@ -255,4 +254,41 @@ public String toString()
         return "WH:" + getName() + "(" + id + ")";
     }
+
+    // TODO: Implement WebhookMessage
```
No and it won't be done in this PR.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/internal/handle/ApplicationCommandCreateHandler.java:
```diff
@@ -0,0 +1,53 @@
+    @Override
+    protected Long handleInternally(DataObject content)
+    {
+        // {"op":0,"s":7,"t":"APPLICATION_COMMAND_CREATE","d":{"name":"ping","guild_id":"163772719836430337","description":"Test command","id":"820420034450030592","version":"820420034450030593","application_id":"420321485757087746"}}
```
Its nice to have an example, and it doesn't harm anything by being here?

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/application/ApplicationCommandCreateEvent.java:
```diff
@@ -0,0 +1,35 @@
+import javax.annotation.Nullable;
+
+/**
+ * Indicates that a {@link Command} was created.
```
This is received for commands from other applications too, I've adjusted the docs to reflect this.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/Emoji.java:
```diff
@@ -0,0 +1,259 @@
+    @Override
+    public String toString()
+    {
+        return "E:" + name + "(" + id + ")";
```
Well it technically represents the same data, so should be ok.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/interactions/components/ButtonInteraction.java:
```diff
@@ -0,0 +1,100 @@
+     */
+    @Nonnull
+    @CheckReturnValue
+    default RestAction<Void> updateButton(@Nullable Button newButton)
```
If users ask for this, we can add it later or let someone else make that contribution.
